### PR TITLE
SortBooks Editorial

### PR DESCRIPTION
**Pull request by eysbutno**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

lmk if u want me to remove the segtree template, altho I personally think it doesn't really affect readability too much

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
how consistent is this segtree with that stock lazy segtree you put in the lazy segtree module

**Comment by TheGamingMousse** [comment]
> how consistent is this segtree with that stock lazy segtree you put in the lazy segtree module

basically the same except this time it's an actual template

**Comment by SansPapyrus683** [comment]
either use the module's one or replace the module's one sometime with this one

**Comment by TheGamingMousse** [comment]
> either use the module's one or replace the module's one sometime with this one

might eventually replace the rurq module's one, idk

**Review by ryanchou-dev — COMMENTED** [review]
good explanation -- is the code ready for review or will you switch up the segtree tmpl?

**Comment by TheGamingMousse** [comment]
> good explanation -- is the code ready for review or will you switch up the segtree tmpl?

ima rewrite rurq module to include the sample problem w/ the template I'm using prolly

File: solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+all the right endpoints. One key observation to make is that if at our current
+indice $i$ there exists a $j$ where $i < j$ and $w_i \leq w_j$, then we can't
+increase the value of any inversion pairs where the second value's index is
+greater than $j$. Thus, we only need to worry about updating the inversion pair
+values on the range $[i + 1, j - 1]$. To find said value of $j$, we utilize a
+monotonic stack.
+
+The most straightforward way to solve the problem from here is to use a lazy
+segment tree to keep track of the best inversion pair value for each indice
+(where the value of an inversion pair $(i, j)$ is stored at index $j$)
```
[the remark this answers is not in the document]
Comment (TheGamingMousse) [reply solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx:35]: good point

**Comment by TheGamingMousse** [comment]
it is worth noting that  there is a BIT solution that doesn't need segtree which might be a bit more straightforward

I'm pretty sure it relies on the same principles that my solution uses but does some funny stuff w/ the monotonic stack

**Comment by SansPapyrus683** [comment]
how "funny" are we talking

**Comment by TheGamingMousse** [comment]
> how "funny" are we talking

as in I still don't exactly know how it works, but can probably figure it out if I really try

**Comment by TheGamingMousse** [comment]
i should probably go figure it out lol

File: solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	vector<vector<array<int, 3>>> tests(n);
+	for (int i = 0; i < q; i++) {
+		int l, r, k;
+		cin >> l >> r >> k;
+		--l, --r;
+		tests[l].push_back({r, k, i});
+	}
+
+	vector<bool> res(q);
+	stack<int> stk;
```
Comment (ryanchou-dev) [inline solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx:167]: stk might not be the best name 👀

**Review by ryanchou-dev — APPROVED** [review]
👌

File: solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx (hunk 2 above)
Comment (TheGamingMousse) [reply solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx:167]: what name would you recommend

File: solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+**Time Complexity:** $\mathcal{O}((N + Q)\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+template <class Info, class Tag> class LazySegtree {
```
Comment (SansPapyrus683) [inline solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx:55]: ok, _now_ is this consistent w the rurq one
if so you can codesnip this

File: solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+monotonic stack.
+
+The most straightforward way to solve the problem from here is to use a lazy
+segment tree to keep track of the best inversion pair value for each index
+(where the value of an inversion pair $(i, j)$ is stored at index $j$)
+and the best inversion pair value on a range. To implement this, notice how
+every inversion pair value stored at an index $j$ is $w_j + x$, where $x$ is
+some value $w_i$. Thus, we can use a range update to apply the best value of $x$.
+
+Note that a more elegant (although less intuitive) Binary Indexed Tree solution
```
Comment (SansPapyrus683) [inline solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx:40]:
```suggestion
Note that a more elegant (although less intuitive) solution with BITs
```

File: solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	for (int i = n - 1; i >= 0; i--) {
+		// maintain a monotonic stack to find the first index j where w[i] >= w[j]
+		while (!indices.empty() && w[indices.top()] < w[i]) { indices.pop(); }
+		int b = indices.empty() ? n : indices.top();
+		indices.push(i);
+		st.upd(i + 1, b - 1, {w[i]});
+		for (auto [r, k, idx] : tests[i]) { res[idx] = st.qry(i, r).applied <= k; }
+	}
+
+	for (int i = 0; i < q; i++) { cout << res[i] << "\n"; }
```
Comment (SansPapyrus683) [inline solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx:177]:
```suggestion
	for (int i = 0; i < q; i++) { cout << res[i] << '\n'; }

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx:56]: done

**Review by SansPapyrus683 — APPROVED** [review]

**Pull request merged**